README.md

# Cartwright
![Tests](https://github.com/jataware/cartwright/actions/workflows/tests.yml/badge.svg)

<img src="docs/assets/cartwright-logo.png" width="175px" align="left"/>
Cartwright is a data profiler that identifies and categorizes spatial and temporal features. Cartwright uses deep learning, natural language processing, and a variety of heuristics to determine whether a column in a dataset contains spatial or temporal information and, if so, what is specifically contained.

Cartwright was built to automate complex data pipelines for heterogenous climate and geopolitical data that are generally oriented around geospatial and temporal features (_think maps and time series_). The challenge that Cartwright solves is automatically detecting those features so they can be parsed and normalized. This problem turns out to be quite tricky, but Cartwright makes it simple.

Cartwright can easily detect things like `country`, `day`, `latitude`, and many other location and time types. Check out Cartwright's [supported categories](https://jataware.github.io/cartwright/categories.html) for a complete listing!

Cartwright is easy to install and works with pretty much any tabular data. It's easy to add new categories too! Learn more about the methodology behind Cartwright, its API, and how to contribute in our [docs](https://jataware.github.io/cartwright).

## Installation

You can install Cartwright from PyPi with `pip install cartwright`.

## Using Cartwright

Imagine we have the following weather dataset:

| x_value  |  y_value   | recorded_at | rainfall |
|:---------|:----------:|------------:|--------|
| 7.942658 | 107.240322 | 07/14/1992  | .2     |
| 7.943745 | 137.240633 | 07/15/1992  | .1     |
| 7.943725 | 139.240664 | 07/16/1992  | .3     |


To the human observer, it's pretty obvious that `x_value` is the longitude column, `y_value` the latitude, `recorded_at` the date, and `rainfall` the actual weather measurement. However, if we're trying to automatically ingest this data into a weather model, we would benefit from knowing this _without human observation_. Enter Cartwright:

```    
from pprint import pprint
from cartwright import categorize

cartwright = categorize.CartwrightClassify()
categories = cartwright.categorize(path="path/to/data.csv")

pprint(categories, sort_dicts=False)
```    

We've now categoriezed each column in this dataset and have automatically determined which column represents latitude, longitude and date. We've also learned the time format (`%m/%d/%Y`) of the date feature.

```
{'x_value': {'category': <Category.geo: 'geo'>,
             'subcategory': <Subcategory.longitude: 'longitude'>,
             'format': None},
 'y_value': {'category': <Category.geo: 'geo'>,
             'subcategory': <Subcategory.latitude: 'latitude'>,
             'format': None},
 'recorded_at': {'category': <Category.time: 'time'>,
                'subcategory': <Subcategory.date: 'date'>,
                'format': '%m/%d/%Y'}}
```

With this information we can now convert the date values to a timestamp and plot a timeseries with other features.

## Resolution Detection

In addition to its ability to categorize spatial and temporal features, Cartwright can determine their resolution. For example, given a dataset like:

```
date,temperature(C)
2019-01-01 00:00:00, 10.2
2019-01-01 02:00:00, 11.7
2019-01-01 04:00:00, 12.3
...
2019-12-31 22:00:00, 10.1
```

Cartwright can detect it's temporal resolution:

```
Resolution(
    uniformity=Uniformity.PERFECT,
    unit=TimeUnit.HOUR,
    resolution=2.0,
    error=0.0,
)
```

For gridded data, which is common in the scientific domain, Cartwright can also determine the spatial resolution (grid size). Check out the docs to learn more about using Cartwright to detect [temporal resolution](https://jataware.github.io/cartwright/temporal_resolution.html) and [spatial resolution](https://jataware.github.io/cartwright/geospatial_resolution.html).

CHANGELOG.md
## 0.0.1 (2022-10-10)
  First release on Pypi
## 0.0.2 (2022-11-16)
  Second release on PyPi.
## 0.0.3 (2022-11-16)
  Third release on PyPi.

cartwright/utils.py
from collections import defaultdict
from fuzzywuzzy import fuzz

import dateutil
import dateutil.parser
import datetime
import logging


def fuzzy_match(word1, word2, ratio_=95):
    Ratio = fuzz.ratio(word1.lower(), word2.lower())

    if Ratio > ratio_:
        return True


def fuzzy_ratio(word1, word2, ratio_=95):
    Ratio = fuzz.ratio(word1.lower(), word2.lower())
    if Ratio > ratio_:
        return True, Ratio


def build_return_date_object(format, util="Util"):
    return {
        "category": "time",
        "subcategory": "date",
        "format": format,
        "match_type": ["LSTM"],
        "Parser": f"{util}",
    }


def build_return_standard_object(category, subcategory, match_type):
    return {
        "category": category,
        "subcategory": subcategory,
        "format": None,
        "match_type": [match_type],
        "Parser": None,
    }


def build_return_timespan(format):
    return {
        "category": "time",
        "subcategory": "timespan",
        "format": format,
        "match_type": ["LSTM"],
        "Parser": None,
    }


def date_util_span(dates):
    validDate = []
    for date in dates:
        dateUtil = dateutil.parser.parse(str(date), dayfirst=False)
        if isinstance(dateUtil, datetime.date):
            validDate.append({"value": date, "standard": dateUtil})
    return validDate


character_tokins = defaultdict(
    int,
    {
        "PAD": 0,
        "UNK": 1,
        "a": 2,
        "b": 3,
        "c": 4,
        "d": 5,
        "e": 6,
        "f": 7,
        "g": 8,
        "h": 9,
        "i": 10,
        "j": 11,
        "k": 12,
        "l": 13,
        "m": 14,
        "n": 15,
        "o": 16,
        "p": 17,
        "q": 18,
        "r": 19,
        "s": 20,
        "t": 21,
        "u": 22,
        "v": 23,
        "w": 24,
        "x": 25,
        "y": 26,
        "z": 27,
        "A": 28,
        "B": 29,
        "C": 30,
        "D": 31,
        "E": 32,
        "F": 33,
        "G": 34,
        "H": 35,
        "I": 36,
        "J": 37,
        "K": 38,
        "L": 39,
        "N": 40,
        "O": 41,
        "P": 42,
        "Q": 43,
        "R": 44,
        "S": 45,
        "T": 46,
        "U": 47,
        "V": 48,
        "W": 49,
        "X": 50,
        "Y": 51,
        "Z": 52,
        "1": 53,
        "2": 54,
        "3": 55,
        "4": 56,
        "5": 57,
        "6": 58,
        "7": 59,
        "8": 60,
        "9": 61,
        "0": 62,
        "'": 63,
        ",": 64,
        ".": 65,
        ";": 66,
        "*": 67,
        "!": 68,
        "@": 68,
        "#": 70,
        "$": 71,
        "%": 72,
        "^": 73,
        "&": 74,
        "(": 75,
        ")": 76,
        "_": 77,
        "=": 78,
        "-": 79,
        ":": 80,
        "+": 81,
        "/": 82,
        "\\": 83,
        "`": 84,
        "~": 85,
        "|": 86,
        "}": 87,
        "{": 88,
    },
)

days_of_the_week_A = [
    "monday",
    "tuesday",
    "wednesday",
    "thursday",
    "friday",
    "saturday",
    "sunday",
]
days_of_the_week_a = ["mon", "tues", "wed", "thur", "fri", "sat", "sun"]
months_of_the_year_B = [
    "january",
    "february",
    "march",
    "april",
    "may",
    "june",
    "july",
    "august",
    "september",
    "october",
    "november",
    "december",
]
months_of_the_year_b = [
    "jan",
    "feb",
    "mar",
    "apr",
    "may",
    "jun",
    "jul",
    "aug",
    "sept",
    "oct",
    "nov",
    "dec",
]
columns_to_classify_and_skip_if_found = [
    {"Lat": "latitude"},
    {"Latitude": "latitude"},
    {"lng": "latitude"},
    {"lon": "longitude"},
    {"long": "longitude"},
    {"Longitude": "longitude"},
    {"ISO2": "ISO2"},
    {"ISO3": "ISO3"},
]
columns_to_classify_if_found = [
    {"Date": "Date"},
    {"Datetime": "Datetime"},
    {"Timestamp": "Timestamp"},
    {"Epoch": "Epoch"},
    {"Time": "Time"},
    {"Year": "Year"},
    {"Month": "Month"},
    {"Lat": "Latitude"},
    {"Latitude": "Latitude"},
    {"lng": "Latitude"},
    {"lon": "Longitude"},
    {"long": "Longitude"},
    {"Longitude": "Longitude"},
    {"Geo": "Geo"},
    {"Coordinates": "Coordinates"},
    {"Location": "Location"},
    {"West": "West"},
    {"South": "South"},
    {"East": "East"},
    {"North": "North"},
    {"Country": "Country"},
    {"CountryName": "CountryName"},
    {"CC": "CC"},
    {"CountryCode": "CountryCode"},
    {"State": "State"},
    {"City": "City"},
    {"Town": "Town"},
    {"Region": "Region"},
    {"Province": "Province"},
    {"Territory": "Territory"},
    {"Address": "Address"},
    {"ISO2": "ISO2"},
    {"ISO3": "ISO3"},
    {"ISO_code": "ISO_code"},
    {"Results": "Results"},
]
